### PR TITLE
New notification provider: SMS Partner API

### PR DESCRIPTION
**Pull request by nicolasverlhiac**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

This adds the support for [SmsPartner.fr](https://www.smspartner.fr/) in order to send SMS notifications in 200+ countries.

## Type of change

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

CleanShot 2024-05-21 at 15 27 45@2x

Event | Before | After
-- | -- | --
UP | - |  CleanShot 2024-05-21 at 15 27 30@2x
DOWN | - |  CleanShot 2024-05-21 at 15 27 27@2x
Certificate-expiry | - |  -
Testing | - |  CleanShot 2024-05-21 at 15 27 33@2x

### CODE REVIEW

**Comment by CommanderStorm** [comment]
I think you are waiting for a discussion.
We only require those for new features.

The process on how to add new notification providers is listed here: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#:~:text=new%20notification%20providers

**Comment by nicolasverlhiac** [comment]
@CommanderStorm Thank you for pointing me in the right direction.
I will look into the guidelines for adding new notification providers at the provided link.

File: server/notification-providers/smspartner.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * @inheritdoc
+     */
+    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+        const okMsg = "Sent Successfully.";
+        const url = "https://api.smspartner.fr/v1/send";
+
+        try {
+            //Lets remove non ascii char
+            let cleanMsg = msg.replace(/[^\x00-\x7F]/g, "");
```
Comment (CommanderStorm) [inline server/notification-providers/smspartner.js:16]: Lets do all `msg`-saintisation here.

```suggestion
            // smspartner does not support non ascii characters and only a maximum 639 characters
            let cleanMsg = msg.replace(/[^\x00-\x7F]/g, "").substring(0, 639);
```

File: server/notification-providers/smspartner.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+            let data = {
+                "apiKey": notification.smspartnerApikey,
+                "gamme": 1,
+                "sender": notification.smspartnerSenderName.substring(0, 11),
+                "phoneNumbers": notification.smspartnerPhoneNumber,
+                "message": cleanMsg.substring(0, 639)
```
Comment (CommanderStorm) [inline server/notification-providers/smspartner.js:23]: see above

```suggestion
                "message": cleanMsg,
```

File: server/notification-providers/smspartner.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                    "cache-control": "no-cache",
+                    "Accept": "application/json",
+                }
+            };
+
+            let resp = await axios.post(url, data, config);
+
+            if (resp.data.success !== true) {
+                let error = "Something gone wrong. Api returned " + resp.data.response.status + ".";
+                this.throwGeneralAxiosError(error);
```
Comment (CommanderStorm) [inline server/notification-providers/smspartner.js:38]: nit: can be shortened.

Also lets not use `throwGeneralAxiosError` twice, as that would lead to `Error: Error:` in the output ^^

```suggestion
                throw Error(`Api returned ${resp.data.response.status}.`);
```

File: src/components/notifications/SMSPartner.vue [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    <div class="mb-3">
+        <label for="smspartner-key" class="form-label">{{ $t("smspartnerApikey") }}</label>
+        <HiddenInput id="smspartner-key" v-model="$parent.notification.smspartnerApikey" :required="true" autocomplete="new-password"></HiddenInput>
+        <div class="form-text">
+            {{ $t("smspartnerApiurl") }}
+            <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
+        </div>
+    </div>
+    <div class="mb-3">
+        <label for="smspartner-phone-number" class="form-label">{{ $t("smspartnerPhoneNumber") }}</label>
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:11]: please extract the part from `smspartnerPhoneNumber` which is a helptext as a helptext ^^
Maybe reword a little to `multiple numbers must be separated by ","`.

> [!TIP]
> You can use https://vue-i18n.intlify.dev/guide/advanced/component.html to use `<code>,</code>` in the helptext

File: src/components/notifications/SMSPartner.vue [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<template>
+    <div class="mb-3">
+        <label for="smspartner-key" class="form-label">{{ $t("smspartnerApikey") }}</label>
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:3]: For common translations (such as this one), we try to stick to the actual words as keys to deduplicate them => reducing the translation effort.

```suggestion
        <label for="smspartner-key" class="form-label">{{ $t("API Key") }}</label>
```

File: src/lang/en.json [hunk 6]
```diff
@@ -753,6 +753,11 @@
     "smseagleUrl": "Your SMSEagle device URL",
     "smseagleEncoding": "Send as Unicode",
     "smseaglePriority": "Message priority (0-9, default = 0)",
+    "smspartnerApikey": "API Key",
```
Comment (CommanderStorm) [inline src/lang/en.json:756]:
```suggestion
```

File: src/lang/en.json [hunk 7]
```diff
@@ -753,6 +753,11 @@
...
+    "smspartnerApiurl": "Find your API key in your dashboard: ",
+    "smspartnerPhoneNumber": "Phone number(s) (multiple numbers, separate by `,`)",
```
Comment (CommanderStorm) [inline src/lang/en.json:758]:
```suggestion
    "smspartnerPhoneNumber": "Phone number(s)",
    "multiple numbers must be separated by": "multiple numbers must be separated by {0}",
```

File: src/lang/en.json [hunk 8]
```diff
@@ -753,6 +753,11 @@
...
+    "smspartnerSenderName": "SMS Sender Name",
+    "smspartnerSenderNameInfo": "Must be between 3 and 11 inclusive and must not contain any special characters.",
```
Comment (CommanderStorm) [inline src/lang/en.json:760]: what do you think about this alternative?

```suggestion
    "smspartnerSenderNameInfo": "Must be between 3..=11 regular characters",
```

File: src/components/notifications/SMSPartner.vue [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        <HiddenInput id="smspartner-key" v-model="$parent.notification.smspartnerApikey" :required="true" autocomplete="new-password"></HiddenInput>
+        <div class="form-text">
+            {{ $t("smspartnerApiurl") }}
+            <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
+        </div>
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:8]: Lets use component translation instead to allow for better translations.

```suggestion
        <i18n-t keypath="smspartnerApiurl" as="div" class="form-text">
            <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
        </i18n-t>
```

File: src/lang/en.json [hunk 10]
```diff
@@ -753,6 +753,11 @@
     "smseagleUrl": "Your SMSEagle device URL",
     "smseagleEncoding": "Send as Unicode",
     "smseaglePriority": "Message priority (0-9, default = 0)",
+    "smspartnerApikey": "API Key",
+    "smspartnerApiurl": "Find your API key in your dashboard: ",
```
Comment (CommanderStorm) [inline src/lang/en.json:757]: See above

```suggestion
    "smspartnerApiurl": "You can find your API key in your dashboard at {0}",
```

File: src/components/notifications/SMSPartner.vue [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    </div>
+    <div class="mb-3">
+        <label for="smspartner-phone-number" class="form-label">{{ $t("smspartnerPhoneNumber") }}</label>
+        <input id="smspartner-phone-number" v-model="$parent.notification.smspartnerPhoneNumber" type="text" class="form-control" required>
+    </div>
+    <div class="mb-3">
+        <label for="smspartner-sender-name" class="form-label">{{ $t("smspartnerSenderName") }}</label>
+        <input id="smspartner-sender-name" v-model="$parent.notification.smspartnerSenderName" type="text" minlength="3" maxlength="11" class="form-control">
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:16]: when you say that only regular characters are allowed. Can you specify this as a [regex input pattern](https://www.w3schools.com/tags/att_input_pattern.asp)?

Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:16]: this input is not set as `required`. Is this intentional?

File: server/notification-providers/smspartner.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        const okMsg = "Sent Successfully.";
+        const url = "https://api.smspartner.fr/v1/send";
+
+        try {
+            //Lets remove non ascii char
+            let cleanMsg = msg.replace(/[^\x00-\x7F]/g, "");
+
+            let data = {
+                "apiKey": notification.smspartnerApikey,
+                "gamme": 1,
```
Comment (CommanderStorm) [inline server/notification-providers/smspartner.js:20]: what is `gamme`?

File: src/components/notifications/SMSPartner.vue [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        <label for="smspartner-key" class="form-label">{{ $t("smspartnerApikey") }}</label>
+        <HiddenInput id="smspartner-key" v-model="$parent.notification.smspartnerApikey" :required="true" autocomplete="new-password"></HiddenInput>
+        <div class="form-text">
+            {{ $t("smspartnerApiurl") }}
+            <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
+        </div>
+    </div>
+    <div class="mb-3">
+        <label for="smspartner-phone-number" class="form-label">{{ $t("smspartnerPhoneNumber") }}</label>
+        <input id="smspartner-phone-number" v-model="$parent.notification.smspartnerPhoneNumber" type="text" class="form-control" required>
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:12]: Can/should we restrict what can be entered into this input further?
What is the allowed format of phone numbers according to their docs?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Looks overall good.
I have left a few UX/i18n-comments below

File: server/notification-providers/smspartner.js (hunk 12 above)
Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js:20]: This is the SMS range, its value must be:
1 for Premium SMS
2 for Low Cost SMS
If this parameter is omitted, SMS messages will be sent in the Premium range.
I think this can be removed.

File: server/notification-providers/smspartner.js (hunk 1 above)
Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js:16]: Thanks, it's much clearer this way.

File: server/notification-providers/smspartner.js (hunk 3 above)
Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js:38]: Better, thank you ! :)

File: src/components/notifications/SMSPartner.vue (hunk 11 above)
Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue:15]: Thank you for your vigilance, indeed, this is not normal, it's an omission 🫣

Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue:15]: Thanks, it's much better with it! :)

File: src/components/notifications/SMSPartner.vue (hunk 4 above)
Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue:12]: Done ! Thank you :)

File: src/components/notifications/SMSPartner.vue (hunk 13 above)
Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue:13]: In the doc, numbers can be :
- in national (06xxxxxxxx) or international (+336xxxxxxxx) format, for French numbers.
- international format (+496xxxxxxxx), for numbers outside France.

Maybe we can use a pattern like this one :
```
minlength="3" maxlength="20" pattern="^[\d+,]+$"
```

**Review by nicolasverlhiac — COMMENTED** [review]
Thank you so much @CommanderStorm for your review, it's much appreciated! :)

File: server/notification-providers/smspartner.js (hunk 12 above)
Comment (CommanderStorm) [reply server/notification-providers/smspartner.js:20]: What is "premium" SMS? 🤔

File: src/components/notifications/SMSPartner.vue (hunk 13 above)
Comment (CommanderStorm) [reply src/components/notifications/SMSPartner.vue:13]: Sounds like a good regex

Comment (CommanderStorm) [reply src/components/notifications/SMSPartner.vue:13]: Do you think adding a help text that national French numbers and international numbers are allowed is beneficial?

File: server/notification-providers/smspartner.js (hunk 12 above)
Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js:20]: The "Premium" range is SMS that can be sent 24 hours a day, while eco can only be sent from 8am to 8pm. But SMS Partner no longer seems to make the ECO range. So the Premium range is induced by default.

File: src/components/notifications/SMSPartner.vue (hunk 13 above)
Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue:13]: I think the "smallest common denominator" is the number in their international format. Maybe just indicate that this is the form in which it should be supplied (international format). I commit in this way. Thank you :)

Comment (nicolasverlhiac) [reply src/components/notifications/SMSPartner.vue]: Done in [389353b](https://github.com/louislam/uptime-kuma/pull/4769/commits/389353b4639151012d4d7afb8de6dbec0bc9240c) :)

File: server/notification-providers/smspartner.js (hunk 12 above)
Comment (CommanderStorm) [reply server/notification-providers/smspartner.js]: ***ECO RANGE DOES WHAT???***

Would it not be better if we send in the eco range from 7:30am to 7:30am and in the premium range otherwise?

Comment (nicolas-loud) [reply server/notification-providers/smspartner.js]: It could have been a great idea! But unfortunately **eco** range no longer exists at SMS Partner :/

Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js]: It could have been a great idea! But unfortunately eco range no longer exists at SMS Partner :/
CleanShot 2024-05-22 at 16 18 47@2x

Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js]: >  As of August 1, 2019, automated sending from 10-digit mobile numbers is no longer permitted on the platform (In accordance with Decision no. 2018-0881 of July 24, 2018).
“In order to strengthen users' confidence in the calls and messages they receive and to enable operators to better manage the flow of calls and SMS/MMS messages sent by automated systems, this decision provides that as of August 1, 2019, certain territorialized numbers (geographic, non-geographic and 10-digit mobile numbers) may no longer be used as caller IDs or message senders presented to the called party for calls or messages sent by automated systems.”

Comment (CommanderStorm) [reply server/notification-providers/smspartner.js]: So this parameter is a no-op? => then lets remove it, or am I missing something?

Comment (nicolasverlhiac) [reply server/notification-providers/smspartner.js]: Yes, I removed this param here : https://github.com/louislam/uptime-kuma/pull/4769/commits/f81cd4411627892eef60a48b6811f643dd97eebb

Sorry for the misunderstanding

**Review by Felitendo — APPROVED** [review]

File: src/lang/en.json [hunk 14]
```diff
@@ -753,11 +753,11 @@
     "smseagleUrl": "Your SMSEagle device URL",
     "smseagleEncoding": "Send as Unicode",
     "smseaglePriority": "Message priority (0-9, default = 0)",
-    "smspartnerApikey": "API Key",
-    "smspartnerApiurl": "Find your API key in your dashboard: ",
-    "smspartnerPhoneNumber": "Phone number(s) (multiple numbers, separate by `,`)",
+    "smspartnerApiurl": "You can find your API key in your dashboard at {0}",
+    "smspartnerPhoneNumber": "Phone number(s)",
+    "multiple numbers must be separated by": "The number must be in the international format {0}, {1}. Multiple numbers must be separated by {2}",
```
Comment (CommanderStorm) [inline src/lang/en.json:758]:
```suggestion
    "smspartnerPhoneNumberHelptext": "The number must be in the international format {0}, {1}. Multiple numbers must be separated by {2}",
```

**Review by CommanderStorm — COMMENTED** [review]
- Thinking about it: lets use a scoped key. The english translation for `multiple numbers must be separated by` is too specific to be usefull for other monitors ^^

File: src/components/notifications/SMSPartner.vue [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+            <i18n-t keypath="smspartnerApiurl" as="div" class="form-text">
+                <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
+            </i18n-t>
+        </div>
+    </div>
+    <div class="mb-3">
+        <label for="smspartner-phone-number" class="form-label">{{ $t("smspartnerPhoneNumber") }}</label>
+        <input id="smspartner-phone-number" v-model="$parent.notification.smspartnerPhoneNumber" type="text" minlength="3" maxlength="20" pattern="^[\d+,]+$" class="form-control" required>
+        <div class="form-text">
+            <i18n-t keypath="multiple numbers must be separated by" as="div" class="form-text">
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:15]: Fuckups happen, should forgot to add this comment ^^

```suggestion
            <i18n-t keypath="smspartnerPhoneNumberHelptext" as="div" class="form-text">
```

File: src/components/notifications/SMSPartner.vue [hunk 16]
```diff
@@ -0,0 +1,39 @@
+<template>
+    <div class="mb-3">
+        <label for="smspartner-key" class="form-label">{{ $t("API Key") }}</label>
+        <HiddenInput id="smspartner-key" v-model="$parent.notification.smspartnerApikey" :required="true" autocomplete="new-password"></HiddenInput>
+        <div class="form-text">
+            <i18n-t keypath="smspartnerApiurl" as="div" class="form-text">
+                <a href="https://my.smspartner.fr/dashboard/api" target="_blank">https://my.smspartner.fr/dashboard/api</a>
```
Comment (CommanderStorm) [inline src/components/notifications/SMSPartner.vue:7]: bit personal preference, but feel like this can be shortened ^^

```suggestion
                <a href="https://my.smspartner.fr/dashboard/api" target="_blank">my.smspartner.fr/dashboard/api</a>
```

**Review by CommanderStorm — APPROVED** [review]
Thanks for the new notification provider! 🎉 



> [!NOTE]
> This PR is part of the v2.0 merge window => see #4500 for the bugs that need to be adressed before we can shit this ^^


<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>
